RemoteLayerTreeDrawingAreaProxy::CommitLayerTree can be processed before WillCommitLayerTree.

mattwoodrow opened the pull request:
#### be985750e45c1fc613c9129570ab23d1896ee983
<pre>
RemoteLayerTreeDrawingAreaProxy::CommitLayerTree can be processed before WillCommitLayerTree.
<a href="https://bugs.webkit.org/show_bug.cgi?id=301392">https://bugs.webkit.org/show_bug.cgi?id=301392</a>
&lt;<a href="https://rdar.apple.com/problem/163307734">rdar://problem/163307734</a>&gt;

Reviewed by Gerald Squelart.

This happens because we use waitForAndDispatchImmediately which caused that
message to be processed synchronously, before others in the queue.

We workaround this by trying to detect willCommitLayerTree messages that have
been superseded, but it&apos;s hard to reason about.

This combines the WillCommitLayerTree and CommitLayerTreeNotTriggered messages
into a single message, and then has waitForDidUpdateActivityState wait on
whichever message is expected next to avoid reordering.

This reverts 266421@main, and replaces it with MESSAGE_CHECK assertions about
the state of commitLayerTreeMessageState and pendingLayerTreeTransactionID.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::waitForMessage):
(IPC::Connection::processIncomingMessage):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::waitForAndDispatchImmediately):
(IPC::Connection::WaitForMessageState::WaitForMessageState):
(IPC::Connection::waitForAsyncReplyAndDispatchImmediately):
(IPC::Connection::waitForMessageForTesting):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeNotTriggered):
(WebKit::RemoteLayerTreeDrawingAreaProxy::willCommitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
(WebKit::RemoteLayerTreeDrawingAreaProxy::waitForDidUpdateActivityState):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::displayDidRefresh):

Canonical link: <a href="https://commits.webkit.org/302461@main">https://commits.webkit.org/302461@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/55bc73979ae6a07caac477027f8b56cd56444d34

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129153 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1411 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136531 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80541 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/22cfd3ef-b870-46f6-a80f-200d6083dc8e) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1344 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1288 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98343 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66218 "") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/aa420728-8c1e-4eaa-89aa-7829f145af21) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132100 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1046 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115693 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78991 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4fd280af-8e34-48f7-bdb5-d5279185bea2) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/968 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79812 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109416 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139005 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1204 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1165 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106877 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1256 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112029 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106712 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27166 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/990 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30553 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53781 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1277 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64630 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1103 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1148 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1201 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -757,7 +757,17 @@ Timeout Connection::timeoutRespectingIgnoreTimeoutsForTesting(Timeout timeout) c
+            return true;
+    }
+    return false;
+}
```
std::find?

cdumez replied:
`std::ranges::find` even (to avoid dealing with iterators). I would have said `std::ranges::contains` but last I check some of our bots didn't like it yet.

squelart commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -909,7 +915,22 @@ void Connection::sendAsyncReply(AsyncReplyID asyncReplyID, Arguments&&... argume
+template<typename T, typename U, typename V> Error Connection::waitForAndDispatchImmediately(uint64_t destinationID, Timeout timeout, OptionSet<WaitForOption> waitForOptions)
+{
+    static_assert(T::canDispatchOutOfOrder, "Can only use waitForAndDispatchImmediately on messages declared with CanDispatchOutOfOrder");
+    auto decoderOrError = waitForMessage(std::array { T::name(), U::name(), V::name() }, destinationID, timeout, waitForOptions);
```
I think these two functions could be combined into one variadic template, for less duplicated source code, something like:
```
template<typename... Ts> Error Connection::waitForAndDispatchImmediately(
[...]
    auto decoderOrError = waitForMessage(std::array { Ts::name()... }, destinationID, timeout, waitForOptions);
[...]
```
You'll need to update the declarations above to match.

This could be done in a future PR, if you preferred to test the waters with this easily-revertible patch first.

squelart commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -481,6 +482,11 @@ class Connection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Connec
+    Error waitForAndDispatchImmediately(const ObjectIdentifierGenericBase<RawValue>& destinationID, Timeout timeout, OptionSet<WaitForOption> waitForOptions = { })
+    {
+        return waitForAndDispatchImmediately<T, U, V>(destinationID.toUInt64(), timeout, waitForOptions);
+    }
```
This could be combined with the above function into a variadic template.

squelart commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -766,7 +766,7 @@ auto Connection::waitForMessage(MessageName messageName, uint64_t destinationID,
     if (signpostsEnabled()) [[unlikely]] {
         signpostIdentifier = generateSignpostIdentifier();
-        WTFBeginSignpost(signpostIdentifier, IPCConnection, "waitForMessage: %" PUBLIC_LOG_STRING, description(messageName).characters());
+        WTFBeginSignpost(signpostIdentifier, IPCConnection, "waitForMessage: %" PUBLIC_LOG_STRING, description(messageNames[0]).characters());
```
I see you're only exposing the first message name, is that intended?
(Probably not *that* critical, but we wouldn't want to confuse our future selves with a non-exhaustive message list.)

squelart commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -803,7 +803,7 @@ auto Connection::waitForMessage(MessageName messageName, uint64_t destinationID,
                 UniqueRef<Decoder>& message = *it;
 
-                if (message->messageName() == messageName && message->destinationID() == destinationID) {
+                if (std::ranges::find(messageNames, message->messageName()) != messageNames.end() && message->destinationID() == destinationID) {
```
Thank you for using ranges::find.
Optional: In this case, since you've already copied the messages names into a `WaitForMessageState`'s `Vector`, could you instead do `waitingForMessage.messageNames.contains(message->messageName())` (similar to what you're already doing at line 1110 below).
`contains()` feels a bit clearer than the long-winded `find() != end()`. :-)

squelart commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -827,7 +827,7 @@ auto Connection::waitForMessage(MessageName messageName, uint64_t destinationID,
         bool wasMessageToWaitForAlreadyDispatched = false;
         m_syncState->dispatchMessages([&](auto nameOfMessageToDispatch, uint64_t destinationOfMessageToDispatch) {
-            wasMessageToWaitForAlreadyDispatched |= messageName == nameOfMessageToDispatch && destinationID == destinationOfMessageToDispatch;
+            wasMessageToWaitForAlreadyDispatched |= std::ranges::find(messageNames, nameOfMessageToDispatch) != messageNames.end() && destinationID == destinationOfMessageToDispatch;
```
Optional: Same as line 806 above, consider `waitingForMessage.messageNames.contains(message->messageName())` for clarity.

squelart reviewed (approved):
r+ as-is (assuming tests are green (or unrelated failures)), but please see my suggestions for now or later.

squelart commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -827,7 +828,7 @@ auto Connection::waitForMessage(MessageName messageName, uint64_t destinationID,
         bool wasMessageToWaitForAlreadyDispatched = false;
         m_syncState->dispatchMessages([&](auto nameOfMessageToDispatch, uint64_t destinationOfMessageToDispatch) {
-            wasMessageToWaitForAlreadyDispatched |= messageName == nameOfMessageToDispatch && destinationID == destinationOfMessageToDispatch;
+            wasMessageToWaitForAlreadyDispatched |=  waitingForMessage.messageNames.contains(nameOfMessageToDispatch) && destinationID == destinationOfMessageToDispatch;
```
Nit: Double space in `|=  waitingForMessage`.

squelart reviewed (commented):
That's a bigger modification of `Connection` than the early versions of this PR, with this added `m_incomingMessagesBlocked` that can really block processing, and its interplay with `m_incomingMessages`.
Could we please get an extra review from an IPC expert, maybe @cdumez ?

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1552,6 +1555,10 @@ void Connection::dispatchSyncStateMessages()
         assertIsCurrent(dispatcher());
         m_syncState->dispatchMessagesAndResetDidScheduleDispatchMessagesForConnection(*this);
     }
+    {
```
Your invariant is:
`hasMessagesFor(this, m_messagesToDispatchWhileWaitingForSyncReply) == m_incomingMessagesBlocked`

between clearing the m_messagesToDispatchWhileWaitingForSyncReply and this toggle, there might be another message that already added to m_messagesToDispatchWhileWaitingForSyncReply. So you end up with m_messagesToDispatchWhileWaitingForSyncReply containing something for this connection and m_incomingMessagesBlocked == false.

I think for clarity, it's better perhaps not to add m_incomingMessagesBlocked that might get confused.
Instead, just keep the m_incomingMessages empty, and check at the message add time if we can use m_incomingMessages or if everything should keep going to  m_messagesToDispatchWhileWaitingForSyncReply

mattwoodrow replied:
The intended invariant was that `m_incomingMessagesBlocked` was true, while there still existed tasks that would run (or are running) `dispatchIncomingMessages` that were scheduled before `m_incomingMessagesBlocked` was set to true.

I think the race is roughly the same, possibly just worded different. If we set `m_incomingMessagesBlocked` to true while dispatching a task with `dispatchSyncStateMessages` but there already exists a scheduled task to run `dispatchSyncStateMessages` (from an earlier sync message), then we'd set it back to false when the first runs, not when the most-recently-scheduled one runs.

Counting, rather than a bool could fix that I think.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -446,7 +446,7 @@ class Connection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Connec
     template<typename T> SendSyncResult<T> sendSync(T&& message, uint64_t destinationID, Timeout = Timeout::infinity(), OptionSet<SendSyncOption> sendSyncOptions = { }); // Main thread only.
 
-    template<typename> Error waitForAndDispatchImmediately(uint64_t destinationID, Timeout, OptionSet<WaitForOption> waitForOptions = { }); // Main thread only.
+    template<typename... Ts> Error waitForAndDispatchImmediately(uint64_t destinationID, Timeout, OptionSet<WaitForOption> waitForOptions = { }); // Main thread only.
```
Would you be willing to omit this change? It's fairly niche use-case, and it brings the burden that we need to support it in both here and stream. And we should test that the implementation is actually correct.

Could we instead use a single message?
E.g. in your patch there's some request that results in either WillCommitLayerTree or CommitLayerTreeNotTriggered.
Can we have something like NotifyLayerTreeCommitStatus(std::optional<TransactionID> willCommit) ?

mattwoodrow replied:
I can, if you'd prefer.

I'll have to combine all three (WillCommitLayerTree, CommitLayerTree, CommitLayerTreeNotTriggered) into a single 'update layer tree state' message. Bit clunky, but happy to have the complexity there rather than IPC.

mattwoodrow committed:
RemoteLayerTreeDrawingAreaProxy::CommitLayerTree can be processed before WillCommitLayerTree.
https://bugs.webkit.org/show_bug.cgi?id=301392
<rdar://problem/163307734>

Reviewed by Gerald Squelart.

This happens because we use waitForAndDispatchImmediately which caused that
message to be processed synchronously, before others in the queue.

We workaround this by trying to detect willCommitLayerTree messages that have
been superseded, but it's hard to reason about.

This combines the WillCommitLayerTree and CommitLayerTreeNotTriggered messages
into a single message, and then has waitForDidUpdateActivityState wait on
whichever message is expected next to avoid reordering.

This reverts 266421@main, and replaces it with MESSAGE_CHECK assertions about
the state of commitLayerTreeMessageState and pendingLayerTreeTransactionID.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::waitForMessage):
(IPC::Connection::processIncomingMessage):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::waitForAndDispatchImmediately):
(IPC::Connection::WaitForMessageState::WaitForMessageState):
(IPC::Connection::waitForAsyncReplyAndDispatchImmediately):
(IPC::Connection::waitForMessageForTesting):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeNotTriggered):
(WebKit::RemoteLayerTreeDrawingAreaProxy::willCommitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
(WebKit::RemoteLayerTreeDrawingAreaProxy::waitForDidUpdateActivityState):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::displayDidRefresh):

Canonical link: https://commits.webkit.org/302461@main
